tests/test_session_handle_open.rs

```diff
@@ -0,0 +1,166 @@
+        if let ServiceEvent::SessionOpen { session_context } = event {
+            self.addr = Some(session_context.address.clone());
+            if session_context.ty.is_outbound() {
+                control.open_protocol(session_context.id, 1.into()).unwrap();
+                self.count += 1;
+            }
+            if self.count >= 10 {
+                control.shutdown().unwrap();
+            }
+        } else if let ServiceEvent::SessionClose { .. } = event {
+            let _ = control.dial(self.addr.clone().unwrap(), TargetProtocol::Single(0.into()));
+        }
+    }
+}
+
+pub fn create<F>(secio: bool, metas: impl Iterator<Item = ProtocolMeta>, shandle: F) -> Service<F>
+where
+    F: ServiceHandle + Unpin,
+{
+    let mut builder = ServiceBuilder::default().forever(true);
+
+    for meta in metas {
+        builder = builder.insert_protocol(meta);
+    }
+
+    if secio {
+        builder
+            .key_pair(SecioKeyPair::secp256k1_generated())
+            .build(shandle)
+    } else {
+        builder.build(shandle)
+    }
+}
+
+fn test_session_handle_open(secio: bool) {
+    let p_handle_1 = PHandle;
+    let s_handle_1 = SHandle {
+        count: 0,
+        addr: None,
+    };
+
+    let p_handle_2 = PHandle;
+    let s_handle_2 = SHandle {
+        count: 0,
+        addr: None,
+    };
+
+    let meta_dummy_1 = MetaBuilder::new()
+        .id(0.into())
+        .service_handle(move || {
+            let handle = Box::new(Dummy);
+            ProtocolHandle::Callback(handle)
+        })
+        .build();
+
+    let meta_dummy_2 = MetaBuilder::new()
+        .id(0.into())
+        .service_handle(move || {
+            let handle = Box::new(Dummy);
+            ProtocolHandle::Callback(handle)
+        })
+        .build();
+
+    let meta_2 = MetaBuilder::new()
+        .id(1.into())
+        .session_handle(move || {
+            let handle = Box::new(p_handle_1.clone());
+            ProtocolHandle::Callback(handle)
+        })
+        .build();
+
+    let meta_2_1 = MetaBuilder::new()
+        .id(1.into())
+        .session_handle(move || {
+            let handle = Box::new(p_handle_2.clone());
+            ProtocolHandle::Callback(handle)
+        })
+        .build();
+
+    let mut service_1 = create(secio, vec![meta_dummy_1, meta_2].into_iter(), s_handle_1);
+    let mut service_2 = create(secio, vec![meta_dummy_2, meta_2_1].into_iter(), s_handle_2);
+
+    let (addr_sender, addr_receiver) = channel::<Multiaddr>();
+    thread::spawn(move || {
+        let mut rt = tokio::runtime::Runtime::new().unwrap();
+        rt.block_on(async move {
+            let listen_addr = service_2
+                .listen("/ip4/127.0.0.1/tcp/0".parse().unwrap())
+                .await
+                .unwrap();
+
+            addr_sender.send(listen_addr).unwrap();
+
+            loop {
+                if service_2.next().await.is_none() {
+                    break;
+                }
+            }
+        });
```

Comment:
suggest adding `assert(10, count)`

Reply:
You cannot confirm which service has reached 10 unless you confirm the direction of the dial

Reply:
`service_1` dial `service_2`, so `service_1` is inbound to `service_2`.

Reply:
> `service_1` dial `service_2`, so `service_1` is inbound to `service_2`.

But in subsequent dials, it is whoever receives the shutdown task that tries to dial to the other

Reply:
You can assert the sum

Reply:
I don't think that's the point, without reach to 10, the service can't automatically shutdown